components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/username/NotificationUsernameRecoveryManager.java

```diff
@@ -198,33 +199,61 @@ private String getUsernameByClaims(UserClaim[] claims, int tenantId)
             UserClaim claim = claims[i];
             if (claim.getClaimURI() != null && claim.getClaimValue() != null) {
 
+                if (log.isDebugEnabled()) {
+                    log.debug("Searching users for " + claim.getClaimURI() + ":" + claim.getClaimValue());
+                }
                 String[] userList = getUserList(tenantId, claim.getClaimURI(),
                         claim.getClaimValue());
 
                 if (userList != null && userList.length > 0) {
-                    if (userList.length == 1) {
-                        return userList[0];
-                    } else {
-                        //If more than one user find the first matching user. Hence need to define unique claims
-                        if (tempUserList != null) {
-                            for (int j = 0; j < tempUserList.length; j++) {
-                                for (int x = 0; x < userList.length; x++) {
-                                    if (tempUserList[j].equals(userList[x])) {
-                                        return userList[x];
-                                    }
+                    if (log.isDebugEnabled()) {
+                        log.debug(" Userlist : " + Arrays.toString(userList));
+                    }
+                    //If more than one user find the first matching user list. Hence need to define unique claims
+                    if (tempUserList != null) {
+                        List<String> users = new ArrayList<String>();
+                        for (int j = 0; j < tempUserList.length; j++) {
```

Comment:
Use foreach loop for better readability like,

`for (String resultedUser : resultedUserList) {`